pints/_optimisers/_hager_zhang_line_search.py

```diff
@@ -0,0 +1,899 @@
+
+    def n_hyper_parameters(self):
+        """ See :meth:`pints.TunableMethod.n_hyper_parameters()`. """
+        return 2
+
+    def running(self):
+        """ See :meth:`Optimiser.running()`. """
+        return self._running
+
+    def set_hyper_parameters(self, x):
+        """
+        See :meth:`pints.TunableMethod.set_hyper_parameters()`.
+
+        The hyper-parameter vector is ``[c1, c2, m]``.
+        ``c1`` is the parameter for the Armijo condition rule,
+        ``0 < c1 < 0.5``.
+        ``c2`` is the parameter for the curvature condition rule,
+        ``c1 < c2 < 1.0``.
+        ``m`` is the number of maximum number of correction matrices
+        that can be stored for the LM-BFGS update.
+        """
+
+        self.__set_wolfe_line_search_parameters(x[0], x[1])
+
+    def __set_wolfe_line_search_parameters(self, c1, c2):
+        """
+        Sets the parameters for the wolfe conditions.
+
+        Parameters
+        ----------
+        c1: float
+        Parameter for the Armijo condition rule, ``0 < c1 < 0.5``.
+
+        c2: float
+        Parameter for the curvature condition rule, ``c1 < c2 < 1.0``.
+        """
+        if(0 < c1 < 0.5 and c1 < c2 < 1.0):
+            self._c1 = c1
+            self._c2 = c2
+        else:
+            cs = self.wolfe_line_search_parameters()
+            raise ValueError('''Invalid wolfe line search parameters!
+                                Parameters must meet the following conditions:
+                                0 < c1 < 0.5 and c1 < c2 < 1.0
+                                currently c1 = ''', cs[0], 'and  c2 = ', cs[1])
+
+    def set_search_direction(self, px):
+        """
+        Sets the search direction vector px.
+        If this is not a descent direction the line searches will fail
+        i.e return nonsense or actually fail
+
+        Parameters
+        ----------
+        px: vector
+        search direction vector this must be of the same length as x
+        """
+        # TODO: maybe add a check that px contains numbers?
+        if(len(px) == self._n_parameters):
+            self.__px = np.asarray(px)
+        else:
+            raise ValueError('Invalid search direction, px must be of length '
+                             + str(self._n_parameters))
+
+    def xbest(self):
+        """ See :meth:`Optimiser.xbest()`. """
+        return self._xbest
+
+    def stop(self):
+        """ See :meth:`Optimiser.stop()`. """
+
+        if self.__wolfe_cond_meet:
+            # wolfe conditions have been meet
+            # or convergence has occurred
+
+            # resetting so optimiser can be recalled
+            self.__wolfe_cond_meet = False
+            self._running = False
+            return 'Wolfe conditions meet'
+        else:
+
+            return False
+
+    def ask(self):
+        """ See :meth:`Optimiser.ask()`. """
+        if np.any(self.__px) is None:
+            raise ValueError(''''Warning search direction has not been set
```

Comment:
I would remove "warning" since it makes it appear as if it's a warning not an error...